### ISSUE TITLE
mpirun training can't return

### ISSUE THREAD

**Issue by cenjiepeng** [yjxiong/temporal-segment-networks#41]
Hi, @yjxiong 

When I use mpirun -np 2 to run the training with 2 GPUs, the program can't return. It stopped at **Optimization Done** like this ( I set the number of iteration to 20 for convenience ),

> I1221 22:17:14.922746 24063 solver.cpp:631] Iteration 0, lr = 0.001
> I1221 22:17:56.933382 24063 solver.cpp:502] Snapshotting to models/hmdb51_split_1_tsn_rgb_bn_inception_iter_20.caffemodel
> I1221 22:17:57.243152 24063 solver.cpp:510] Snapshotting solver state to models/hmdb51_split_1_tsn_rgb_bn_inception_iter_20.solverstate
> I1221 22:17:57.486786 24063 solver.cpp:406] Iteration 20, loss = 3.53084
> I1221 22:17:57.486837 24063 solver.cpp:411] Optimization Done.
> I1221 22:17:57.486843 24063 caffe.cpp:187] Optimization Done.

and never return to system. The training had been done, so i suspect it is the mpirun couldn't return. Then i run with `--mca mpi_common_cuda_verbose 100`  for additional information, and i get this at the end,

> I1221 22:47:31.502188 27430 solver.cpp:406] Iteration 20, loss = 3.50854
> I1221 22:47:31.502244 27430 solver.cpp:411] Optimization Done.
> I1221 22:47:31.502249 27430 caffe.cpp:187] Optimization Done.
> [UBUNTUB:27430] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 2
> [UBUNTUB:27431] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 2
> [UBUNTUB:27430] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 1
> [UBUNTUB:27431] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 1
> [UBUNTUB:27430] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 0
> [UBUNTUB:27431] CUDA: mca_common_cuda_fini, never completed initialization so skipping fini, ref_count is now 0

which i don't understand. And I have tried launching the training without mpirun ( i.e run caffe train --slover=... directly with one GPU ), the program returns normally, 

> I1221 22:55:10.266736 28827 solver.cpp:510] Snapshotting solver state to models/hmdb51_split_1_tsn_rgb_bn_inception_iter_20.solverstate
> I1221 22:55:10.510454 28827 solver.cpp:406] Iteration 20, loss = 3.61723
> I1221 22:55:10.510510 28827 solver.cpp:411] Optimization Done.
> I1221 22:55:10.510516 28827 caffe.cpp:187] Optimization Done.
> Setting device 0
> cenjiepeng@UBUNTUB:~/temporal-segment-networks$ 

I run it on Ubuntu 16.04 + CUDA 8.0 + OpenMPI 1.8.8 and I have tried OpenMPI 1.8.7 and 1.10.2, but it didn't work. Have you ever encouter this problem?

**Comment by yjxiong** [comment]
On some machine, it is possible to observe this. However, once you see  `Optimization Done`, it is fine to use `CTRL+C` to close the program.

**Comment by cenjiepeng** [comment]
Thank you, @yjxiong.
Could you tell me the reason causing it?  And does it any way to fix this? As it brings some bother if the program couldn't be closed automatically.

**Comment by yjxiong** [comment]
We currently don't have a fix because this rarely happens on our machines. You are welcome to contribute a fix to this issue.

**Comment by cenjiepeng** [comment]
Thanks for your help. I will have a try.

**Comment by KnightOfTheMoonlight** [comment]
You can just kill the process. Find all the process ids and then kill them.
